### PR TITLE
feat: Azure Blob Storage service

### PR DESCRIPTION
**Pull request by tbarlow12**
Wrapper service for Azure Blob Storage functionality. To be used in update of deployment process for rollback capability.

Resolves [AB#361]

### CODE REVIEW

**Comment by mydiemho** [comment]
@tbarlow12 please also add to description on what this PR is about

File: src/services/azureBlobStorageService.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    uploadFileToBlockBlob(Aborter.none, path, this.getBlockBlobURL(containerName, name));
+  };
+  
+  /**
+   * Delete a blob from Azure Blob Storage
+   * @param containerName Name of container containing blob
+   * @param blobName Blob to delete
+   */
+  public async deleteFile(containerName: string, blobName: string): Promise<void> {
+    const blockBlobUrl = await this.getBlockBlobURL(containerName, blobName)
```
Comment (PIC123) [inline src/services/azureBlobStorageService.ts]: Can we make this one line like you have done with the deleteContainer method?

Comment (tbarlow12) [reply src/services/azureBlobStorageService.ts]: That's a good question... Do they both get awaited?

File: src/services/azureBlobStorageService.test.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  const containers = MockFactory.createTestAzureContainers();
+  const sls = MockFactory.createTestServerless();
+  const options = MockFactory.createTestServerlessOptions();
+  const blockBlobUrl = MockFactory.createTestBlockBlobUrl(containerName, filePath);
+
+  let service: AzureBlobStorageService;
+
+  beforeAll(() => {
+    BlockBlobURL.fromContainerURL = jest.fn(() => blockBlobUrl) as any;
```
[the remark this answers is not in the document]
Comment (wbreza) [reply src/services/azureBlobStorageService.test.ts]: Block blobs are the default blobs and should be good enough for our use case.  It was good enough for VoTT and should solve all the requirements for our current use cases.

https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs

File: src/services/azureBlobStorageService.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+   * @param ext - Extension of files to filter on when retrieving files
+   * from container
+   */
+  public async listFiles(containerName: string, ext?: string): Promise<string[]> {
+    const result: string[] = [];
+    let marker;
+    const containerURL = this.getContainerURL(containerName);
+    do {
+      const listBlobsResponse = await containerURL.listBlobFlatSegment(
+        Aborter.none,
```
[the remark this answers is not in the document]
Comment (wbreza) [reply src/services/azureBlobStorageService.ts]: i believe marker is used for paging scenarios.  It's undefined on the first iteration so the first page of data is returned.

File: src/services/azureBlobStorageService.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  /**
+   * Lists the containers within the Azure Blob Storage account
+   */
+  public async listContainers() {
+    const result: string[] = [];
+    let marker;
+    do {
+      const listContainersResponse = await this.getServiceURL().listContainersSegment(
+        Aborter.none,
+        marker,
```
[the remark this answers is not in the document]
Comment (wbreza) [reply src/services/azureBlobStorageService.ts]: See above.

File: src/services/azureBlobStorageService.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    const blockBlobUrl = await this.getBlockBlobURL(containerName, blobName)
+    await blockBlobUrl.delete(Aborter.none);
+  }
+
+  /**
+   * Lists files in container
+   * @param ext - Extension of files to filter on when retrieving files
+   * from container
+   */
+  public async listFiles(containerName: string, ext?: string): Promise<string[]> {
```
Comment (wbreza) [inline src/services/azureBlobStorageService.ts]: Please validate input arguments using `Guard` to make sure they satisfy the method requirements.

File: src/services/azureBlobStorageService.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  public async deleteContainer(containerName: string): Promise<void> {
+    const containerUrl = await this.getContainerURL(containerName)
+    await containerUrl.delete(Aborter.none);
+  }
+
+  /**
+   * Get ServiceURL object for Azure Blob Storage Account
+   */
+  private getServiceURL(): ServiceURL {
+    const credential = this.credentials
```
Comment (wbreza) [inline src/services/azureBlobStorageService.ts]: Any reason not to use this.credentials directly vs split out into separate variable?

File: src/services/azureBlobStorageService.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+   * from container
+   */
+  public async listFiles(containerName: string, ext?: string): Promise<string[]> {
+    const result: string[] = [];
+    let marker;
+    const containerURL = this.getContainerURL(containerName);
+    do {
+      const listBlobsResponse = await containerURL.listBlobFlatSegment(
+        Aborter.none,
+        marker,
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/services/azureBlobStorageService.ts]: It's actually an optional param to begin with. Doesn't matter until you've already got a response

[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/services/azureBlobStorageService.ts]: https://github.com/Azure/azure-storage-js/blob/master/blob/README.md#code-samples

File: src/services/azureBlobStorageService.ts (hunk 6 above)
Comment (tbarlow12) [reply src/services/azureBlobStorageService.ts]: Nope :)

File: src/services/azureBlobStorageService.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  /**
+   * Upload a file to Azure Blob Storage
+   * @param path Path of file to upload
+   * @param containerName Name of container in Azure Blob storage for upload
+   * @param blobName Name of blob file created as a result of upload
+   */
+  public async uploadFile(path: string, containerName: string, blobName?: string) {
+    const name = blobName || path.replace(/^.*[\\\/]/, "");
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/services/azureBlobStorageService.ts]: Fixed

**Pull request merged**